### PR TITLE
test: Assert Assistant code block

### PR DESCRIPTION
**Pull request by dcalhoun**
Fixes https://github.com/Automattic/dotcom-forge/issues/7698.

## Proposed Changes

Refactor the Assistant code block to a separate file to enable automated tests.
This is required as we currently mock `react-markdown` due to its [ESM-only
build](https://github.com/Automattic/studio/pull/261#issuecomment-2173693691).

## Testing Instructions

N/A, no user-facing changes.

## Pre-merge Checklist

- [x] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: jest-setup.ts [hunk 1]
```diff
@@ -34,8 +34,4 @@ jest.mock( './src/hooks/use-offline', () => ( {
 	useOffline: jest.fn().mockReturnValue( false ),
 } ) );
 
-jest.mock( 'strip-ansi', () => ( {
-	default: jest.fn().mockImplementation( ( str: string ) => str ),
-} ) );
```
Comment (dcalhoun) [inline jest-setup.ts:39]: Relocated to the `__mocks__` directory to co-locate third-party mocks and make the mock universal for all tests.

File: src/components/assistant-code-block.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	useEffect( () => {
+		if ( blocks ) {
+			const block = blocks?.find( ( block ) => block.codeBlockContent === content );
+			if ( block ) {
+				setCliOutput( block?.cliOutput ? stripAnsi( block.cliOutput ) : null );
+				setCliStatus( block?.cliStatus ?? null );
+				setCliTime( block?.cliTime ?? null );
+			}
+		}
+	}, [ blocks, cliOutput, content, setCliOutput, setCliStatus, setCliTime ] );
```
Comment (dcalhoun) [inline src/components/assistant-code-block.tsx:51]: `blocks` was added addressing the missing dependency lint warning.

File: src/components/assistant-code-block.tsx
Comment (dcalhoun) [inline src/components/assistant-code-block.tsx:1]: I explored enabling Jest's experimental ESM support to remove the `react-markdown` mock (https://github.com/Automattic/studio/commit/1264f3acaa82b0ffdfbcbaf11a8b349c9b35cbd9, https://github.com/Automattic/studio/commit/312100493dd9be69a7dd509f2bd4a90f6c8a5c1d, https://github.com/Automattic/studio/commit/66f9b14631479dd7db2d53299c6f322e665a8e71). I was successful getting Jest tests to pass, but encountered errors building/running the app itself. So, I pivoted to relocating this logic instead to enable more straightforward unit tests.

Comment (fluiddot) [reply src/components/assistant-code-block.tsx:1]: I prefer this disposition of having a separated component. Thanks David for making the change 🙇 !

File: src/components/tests/assistant-code-block.test.tsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		expect( screen.getByText( 'Copy' ) ).toBeInTheDocument();
+	} );
+
+	it( 'should display the "run" button for elligble wp-cli commands without placeholder content', () => {
+		render( <CodeBlock className="language-bash" children="wp --version" /> );
+
+		expect( screen.getByText( 'Run' ) ).toBeInTheDocument();
+	} );
+
+	it( 'should hide the "run" button for inellible non-wp-cli code', () => {
```
Comment (fluiddot) [inline src/components/tests/assistant-code-block.test.tsx]: Not sure if there's a typo in the word `inellible`, should be `ineligible`?

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 !

File: src/components/tests/assistant-code-block.test.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+		expect( screen.queryByText( 'Run' ) ).not.toBeInTheDocument();
+	} );
+
+	it( 'should hide the "run" button for inellible wp-cli commands with placeholder content', () => {
+		render(
+			<CodeBlock className="language-bash" children="wp plugin activate <example-plugin>" />
+		);
+
+		expect( screen.queryByText( 'Run' ) ).not.toBeInTheDocument();
+	} );
```
Comment (fluiddot) [inline src/components/tests/assistant-code-block.test.tsx]: I noticed this test case is the same as the one above (`should hide the "run" button for inellible non-wp-cli code`), I wonder if it was duplicated by mistake.

File: src/components/assistant-code-block.tsx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+export default function createCodeComponent( contextProps: ContextProps ) {
+	return ( props: CodeBlockProps ) => <CodeBlock { ...contextProps } { ...props } />;
+}
+
+function CodeBlock( props: ContextProps & CodeBlockProps ) {
+	const content = String( props.children ).trim();
+	const containsWPCommand = /\bwp\s/.test( content );
+	const wpCommandCount = ( content.match( /\bwp\s/g ) || [] ).length;
+	const containsSingleWPCommand = wpCommandCount === 1;
+	const containsAngleBrackets = /<.*>/.test( content );
```
Comment (fluiddot) [inline src/components/assistant-code-block.tsx:28]: I noticed when testing that AI might respond with different placeholder characters ([reference](https://github.com/Automattic/studio/pull/232#discussion_r1654495584)). I've created [a PR](https://github.com/Automattic/studio/pull/309) to improve this but I wanted to share that I created it branching off from this PR.

File: src/components/tests/assistant-code-block.test.tsx (hunk 3 above)
Comment (dcalhoun) [reply src/components/tests/assistant-code-block.test.tsx]: Thanks for catching this typo. Addressed in https://github.com/Automattic/studio/pull/305/commits/2ae36fd29f28d6ce25b93887c7584a476521f2b7.

File: src/components/tests/assistant-code-block.test.tsx (hunk 4 above)
Comment (dcalhoun) [reply src/components/tests/assistant-code-block.test.tsx]: Thanks for catching this error. I addressed this in https://github.com/Automattic/studio/pull/305/commits/9e759cefeb33f821aaf83c83a2fe26d535b13ac9 by implementing the intended test case of non-`wp-cli` code.

File: src/components/assistant-code-block.tsx (hunk 5 above)
Comment (dcalhoun) [reply src/components/assistant-code-block.tsx:28]: @fluiddot would you prefer merging this as-is or first merge https://github.com/Automattic/studio/pull/309?

Comment (fluiddot) [reply src/components/assistant-code-block.tsx:28]: I don't have a strong preference. Since this PR is already approved, feel free to merge as-is.

**Pull request merged**